### PR TITLE
Implement RemoteExecutor

### PR DESCRIPTION
**Pull request by goedderz**

### CODE REVIEW

**Comment by goedderz** [comment]
jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3222/

File: arangod/Aql/ClusterNodes.cpp [hunk 1]
```diff
@@ -92,7 +93,46 @@ RemoteNode::RemoteNode(ExecutionPlan* plan, arangodb::velocypack::Slice const& b
 /// @brief creates corresponding ExecutionBlock
 std::unique_ptr<ExecutionBlock> RemoteNode::createBlock(
     ExecutionEngine& engine, std::unordered_map<ExecutionNode*, ExecutionBlock*> const&) const {
-  return std::make_unique<RemoteBlock>(&engine, this, server(), ownName(), queryId());
+  // TODO I'm assuming here that input and output registers are the same for
```
Comment (mchacki) [inline arangod/Aql/ClusterNodes.cpp]: Yes they are.
The remote can neither discard nor add something.
Please remove this TODO

File: arangod/Aql/RemoteExecutor.cpp [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+
+  VPackOptions options(VPackOptions::Defaults);
+  options.buildUnindexedArrays = true;
+  options.buildUnindexedObjects = true;
+
+  VPackBuilder builder(&options);
+  builder.openObject();
+
+  // Backwards Compatibility 3.3
+  builder.add("exhausted", VPackValue(false));
```
Comment (mchacki) [inline arangod/Aql/RemoteExecutor.cpp]: Happy to delete this now!

**Comment by goedderz** [comment]
Review comments are implemented, plus the fix from 923b6e8.

New CI run:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3276/

**Comment by goedderz** [comment]
Last run results are gone. http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3310/

**Comment by goedderz** [comment]
Lasts tests were broken, retrying with a fix: http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3327/

File: arangod/Aql/RemoteExecutor.cpp (hunk 2 above)
Comment (goedderz) [reply arangod/Aql/RemoteExecutor.cpp]: @mchacki Removing this broke the cluster tests 😕. So I re-added it, tests are running again.

Comment (mchacki) [reply arangod/Aql/RemoteExecutor.cpp]: =0, i am unhappy about this.
Need to investigate who still checks exhausted.

Comment (goedderz) [reply arangod/Aql/RemoteExecutor.cpp]: I'll open an issue, we'll solve this separately and merge this PR independently - agreed?

**Comment by goedderz** [comment]
One unrelated shutdown failure, despite that the last PR run is green.

File: arangod/Aql/RemoteExecutor.cpp [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    _lastTicketId = 0;
+  }
+  return true;
+}
+
+arangodb::Result ExecutionBlockImpl<RemoteExecutor>::handleCommErrors(ClusterCommResult* res) const {
+  if (res->status == CL_COMM_TIMEOUT || res->status == CL_COMM_BACKEND_UNAVAILABLE) {
+    return {res->getErrorCode(), res->stringifyErrorMessage()};
+  }
+  if (res->status == CL_COMM_ERROR) {
```
Comment (graetzer) [inline arangod/Aql/RemoteExecutor.cpp]: Just so it is not forgotten, I think we should verify this error handling code and align it with the code in ClusterMethods.cpp.

It seems that depening on the codepath used this check might be incorrect, I think this is only ever set if `res->result == nullptr`. At least that seems to be the case in https://github.com/arangodb/arangodb/blob/devel/arangod/Cluster/ClusterComm.h#L247

Comment (goedderz) [reply arangod/Aql/RemoteExecutor.cpp]: I opened an internal ticket so this won't be forgotten.

**Pull request merged**